keel-artifact/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/BaseArtifactSupplier.kt

```diff
@@ -0,0 +1,23 @@
+package com.netflix.spinnaker.keel.artifacts
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactMetadata
+import com.netflix.spinnaker.keel.api.artifacts.PublishedArtifact
+import com.netflix.spinnaker.keel.services.ArtifactMetadataService
+import org.slf4j.LoggerFactory
+
+abstract class BaseArtifactSupplier (
```

Comment:
Consider documenting the class.

Reply:
Actually, I think this class may be unnecessary. It's a bit misleading in that it doesn't implement the `ArtifactSupplier` interface despite the name. It looks like all you wanted to do was to reuse the `getArtifactMetadataInternal` function, so you could define that as an extension function in a kotlin module (not a class), and import that function in the `ArtifactSupplier` implementations. I don't think this is a blocker to merging, but in my understanding it would be a more idiomatic way to do it in kotlin.

Reply:
The other, perhaps better alternative would be to make it an actual base class that implements the interface. Then you could have a default implementation of `getArtifactMetadata` in the abstract class and remove the implementations from all the sub-classes.

Reply:
I tried to implement it and faced an obstacle - my current logic is:
1. whenever a new version of an artifact is being saved, publish an `ArtifactSaved` event
2. in `ArtifactSaved` listener, call `artifactSupplier.getArtifactMetadata(event.artifact)`
3. store this information in the DB.
4. in `ApplicationService` call `repository.getArtifactBuildMetadata(..)` to fetch information, or call fallback.

Changing it to the logic you've proposed for the concrete implementations (debians, npms etc) will not have the option to call the DB to fetch this information since we don't want to include DB calls in those classes. Another option is to store the fallback build and git data in the DB, but I'm not sure what is the right way to do it.
Does it make sense to you? Am I missing something?

Reply:
> Changing it to the logic you've proposed for the concrete implementations (debians, npms etc) will not have the option to call the DB to fetch this information since we don't want to include DB calls in those classes.

Ah, right. I may have mixed those up, sorry. You can still use the idea of the abstract base class for common behavior, though for the event flow, and just keep `ApplicationService` as is.